File: leetcode/easy/1137-tribonacci/dp.py
"""
  @author Vortexx2
  Problem 1137 - Nth Tribonacci Number

  Runtime - 29 ms
  Memory Usage - 13.8 MB
"""


class Solution:
  def tribonacci(self, n: int) -> int:
    self.seq = [0, 1, 1]

    if n <= 2:
      return self.seq[n]

    return self.helper(n)

  def helper(self, n: int) -> int:

    if n < len(self.seq):
      return self.seq[n]

    self.seq.append(self.helper(n - 1) +
                    self.helper(n - 2) + self.helper(n - 3))

    return self.seq[n]

File: leetcode/easy/1137-tribonacci/dp.cpp
/**
 * @file dp.cpp
 * @author Vortexx2
 * @brief Problem 1137 - Nth Tribonacci Number
 * @date 26-06-2022
 * 
 * Runtime - 0 ms
 * Memory Usage - 6.2 MB
 */
#include <iostream>
#include <vector>

using namespace std;

class Solution {
 public:
  vector<int> seq = {0, 1, 1};
  int tribonacci(int n) {
    if (n <= 2) return seq[n];

    return helper(n);
  }

  int helper(int n) {
    if (n < seq.size()) return seq[n];

    seq.push_back(helper(n - 1) + helper(n - 2) + helper(n - 3));
    return seq[n];
  }
};